src/Swarm/Game/Scenario/Status.hs

```diff
@@ -0,0 +1,235 @@
+angleBracket x = "<" <> x <> ">"
+
+instance Show BestRecords where
+  show (BestRecords a b c d) =
+    unlines $
+      map
+        angleBracket
+        [ show a
+        , show b
+        , show c
+        , show d
+        ]
+
+emptyBest :: ZonedTime -> BestRecords
+emptyBest t = BestRecords x x x x
+ where
+  x = Metric Attempted $ ProgressStats t emptyAttemptMetric
+
+updateBest :: ProgressMetric -> BestRecords -> BestRecords
+updateBest newPlayMetric (BestRecords oldA oldB oldC oldD) =
+  BestRecords
+    (bestTime oldA scenarioElapsed)
+    (bestTime oldB scenarioElapsedTicks)
+    (bestSize oldC sourceTextLength)
+    (bestSize oldD astSize)
+ where
+  f x y = chooseBetter y newPlayMetric x
+  bestTime x y = f x (Just . view (scenarioAttemptMetrics . scenarioDurationMetrics . y))
+  bestSize x y = f x (fmap y . view (scenarioAttemptMetrics . scenarioCodeMetrics))
+
+makeLensesWith (lensRules & generateSignatures .~ False) ''BestRecords
+
+-- | The best status of the scenario, measured in real world time.
+scenarioBestByTime :: Lens' BestRecords ProgressMetric
+
+-- | The best status of the scenario, measured in game ticks.
+scenarioBestByTicks :: Lens' BestRecords ProgressMetric
+
+-- | The best code size of the scenario, measured both in character count and AST size.
+scenarioBestByCharCount :: Lens' BestRecords ProgressMetric
+
+-- | The best code size of the scenario, measured both in character count and AST size.
+scenarioBestByAstSize :: Lens' BestRecords ProgressMetric
+
+instance FromJSON BestRecords where
+  parseJSON = genericParseJSON scenarioOptions
+
+instance ToJSON BestRecords where
+  toEncoding = genericToEncoding scenarioOptions
+  toJSON = genericToJSON scenarioOptions
+
+bestToMap :: BestRecords -> Map BestByCriteria ProgressMetric
+bestToMap (BestRecords t1 t2 s1 s2) =
+  M.fromList $ durationElements <> mapMaybe (traverse ensurePresent) codeSizeElements
+ where
+  durationElements =
+    [ (BestByTime, t1)
+    , (BestByTicks, t2)
+    ]
+  codeSizeElements =
+    [ (BestByCharCount, s1)
+    , (BestByAstSize, s2)
+    ]
+
+  ensurePresent x =
+    (getMetric x ^. scenarioAttemptMetrics . scenarioCodeMetrics) >> Just x
+
+-- | Uses the start time of the play-attempt to de-dupe
+-- records that are from the same game. The start time should
+-- be sufficient to uniquely identify a game.
+getBestGroups ::
+  BestRecords ->
+  [(Metric ProgressStats, NonEmpty BestByCriteria)]
+getBestGroups =
+  rearrangeTuples . M.toList . bestToMap
+ where
+  groupByStartTime = NE.groupAllWith $ view scenarioStarted . getMetric . snd
+  rearrangeTuples = map (snd . NE.head &&& NE.map fst) . groupByStartTime
+
+-- * High-level status of scenario play
+
+-- | A "ScenarioStatus" stores the status of a scenario along with
+--   appropriate metadata: "NotStarted", or "Played".
+--   The "Played" status has two sub-states: "Attempted" or "Completed".
+data ScenarioStatus
+  = NotStarted
+  | Played ProgressMetric BestRecords
```

Comment:
This makes for a nice test case to discuss our general approach to breaking changes (there has been some discussion of this before though I cannot find it at the moment).   The constructors of the `ScenarioStatus` type changed, and therefore the JSON format for saving `ScenarioStatus` values has changed. This means that if you have played any scenarios previously, and then update to a version of Swarm with this PR merged, it will break because it will complain that it cannot parse the scenario info files, and the only way to fix it is to blow away `~/.local/share/swarm/saves`.  This is annoying both because users will not know what they need to do, and also because they will necessarily lose all their saved progress.

Since we are just in alpha mode, I am not opposed to breaking things.  However, eventually, once we make a 1.0 release we will want to be more careful about backwards compatibility.  This could be a good opportunity to practice/experiment.

A concrete proposal is to adopt the policy that version x.n can always read save files written by version x.n and x.(n-1) (but it will only write them using version x.n format).  So you can upgrade just by running the new version and your old files will be converted.  In this case we could put in some special code to parse the old scenario info file format and turn it into the new format.  We might want to add version numbers like we already did for scenario description files.

Reply:
Well the “saves” are very alpha, but in the future we could utilize named optional fields or custom FromJSON instances to parse previous versions.

At minimum we could produce a nice failure when encountering the old style save. @kostmo if you want, this should be easy with custom instance and `fail` with message on alternative (`<|>`).

Reply:
> it will break because it will complain that it cannot parse the scenario info files, and the only way to fix it is to blow away `~/.local/share/swarm/saves`.

This is precisely why I turned loading failure from a **fatal** error into a **warning** in #1116.  But, it is true, the saved progress will still be gone.

Reply:
> A concrete proposal is to adopt the policy that version x.n can always read save files written by version x.n and x.(n-1) (but it will only write them using version x.n format). So you can upgrade just by running the new version and your old files will be converted.

I like this strategy, as there is a definite protocol for purging the tech debt of legacy format-parsing code.

For now, though, let's enjoy the privilege of an alpha version and ruthlessly refactor, unburdened by tech debt!  Perhaps the beta release is the time to commission support for legacy format parsing.

Reply:
> At minimum we could produce a nice failure when encountering the old style save.

Filed #1124 for this.